engine/engine_manager.go

```diff
@@ -318,6 +350,48 @@ func (em *Manager) EngineLog(req *rpc.LogRequest, srv rpc.EngineManagerService_E
 	return nil
 }
 
+func (em *Manager) EngineStream(req *empty.Empty, srv rpc.EngineManagerService_EngineStreamServer) error {
+	responseChan := make(chan *rpc.EngineResponse)
+	em.lock.Lock()
+	em.rpcStreams[responseChan] = struct{}{}
+	em.lock.Unlock()
+	defer func() {
+		em.lock.Lock()
+		delete(em.rpcStreams, responseChan)
+		em.lock.Unlock()
+	}()
+
+	var initialEngineState []*rpc.EngineResponse
+	em.lock.RLock()
+	for _, e := range em.engineLaunchers {
+		e.lock.RLock()
+		name := e.currentEngine.EngineName
+		e.lock.RUnlock()
+		response, err := em.processManager.ProcessGet(nil, &rpc.ProcessGetRequest{
+			Name: name,
+		})
+		if err != nil {
+			return err
+		}
+		initialEngineState = append(initialEngineState, e.RPCResponse(response))
+	}
+	em.lock.RUnlock()
+
+	for _, resp := range initialEngineState {
+		if err := srv.Send(resp); err != nil {
+			return err
+		}
+	}
+
+	for resp := range responseChan {
+		if err := srv.Send(resp); err != nil {
```

Comment:
Kubernetes retransmit all the states every 30 seconds as you know just in case the delta got wrong sometimes. We may need to do it too.

Reply:
Yeah, this makes sense. Otherwise the client would have to guarantee being able to update the `object` state without errors or it loses the update, which doesn't make sense realistically.

Reply:
Since the logic for getting a live `Watch` on the `objects` and re-polling the status manually every `x` seconds is different, this is something that'll probably be left to the client (`longhorn-manager`) to implement. We can likely have one method responsible for streaming real-time changes and another responsible for checking for the definitive `list` to represent the status every `x` seconds.